feat: added rtl direction to all of ant-design components

saeedrahimi opened the pull request:
### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / document update
- [x] Component style update
- [ ] TypeScript definition update
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link
close #4051

Previous pull request was on `feature` branch: #18006 
### 💡 Background and solution

As requirement of this [issue](https://github.com/ant-design/ant-design/issues/4051) components of ant-design doesn't support right to left direction, and those components are not useful in some languages like Persian(Farsi) or Arabic.

Solution: 
I started to develop Rtl style on ant-design and as @zombieJ  mentioned in issue i used ConfigProvider to set direction of component.

direction property added to ConfigProvider.
```
<ConfigProvider direction="rtl">
      <Cascader  options={options} placeholder="select an item"/>
</ConfigProvider>
```
in style files added prefixed rtl styles to change direction of displayed items
```
  .@{inputClass}-suffix {
    right: @input-padding-horizontal-base + 1px;

    .@{inputClass}-rtl & {
      right: auto;
      left: @input-padding-horizontal-base + 1px;
    }
  }
```
so when RTL'ed components wrapped by ConfigProvider with `direction='rtl'` property, all child components (also grid-system) will shown in right-to-left direction.
![ant-rtl](https://user-images.githubusercontent.com/12451743/67393053-7bb3b400-f5ae-11e9-82af-361506894684.JPG)

## RTL Progress

- [x] Button
- [x] Icon
- [x] Typography
- [x] Grid
- [x] Layout
- [x] Affix
- [x] Breadcrumb
- [x] Dropdown
- [x] Menu
- [x] Pagination
- [x] PageHeader
- [x] Steps
- [x] AutoComplete
- [x] CheckBox
- [x] Cascader
- [x] DatePicker
- [x] Form
- [x] InputNumber
- [x] Input
- [x] Mention
- [x] Rate
- [x] Radio
- [x] Switch
- [x] Slider
- [x] Select
- [x] TreeSelect
- [x] Transfer
- [x] TimePicker
- [x] Upload
- [x] Avatar
- [x] Badge
- [x] Comment
- [x] Collapse
- [x] Carousel
- [x] Cad
- [x] Calendar
- [x] Description List
- [x] Empty
- [x] List
- [x] Popover
- [x] Statistic
- [x] Tree
- [x] Tooltip
- [x] Timeline
- [x] Tag
- [x] Tabs (Done)
- [x] Table
- [x] Alert
- [x] Drawer
- [x] Modal
- [ ] Message
- [ ] Notification
- [x] Progress
- [x] Popconfirm
- [x] Result
- [x] Spin
- [x] Skeleton
- [x] Anchor (No need to RTL)
- [x] BackTop
- [x] ConfigProvider
- [x] Divider (No need to RTL)
- [x] Front page

### 📝 Changelog

<!--
Describe changes from userside, and list all potential break changes or other risks.
--->

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |           |
| 🇨🇳 Chinese |           |

### ☑️ Self Check before Merge

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed


-----
[View rendered components/config-provider/demo/direction.md](https://github.com/saeedrahimi/ant-design/blob/ant-rtl/components/config-provider/demo/direction.md)
[View rendered components/config-provider/index.en-US.md](https://github.com/saeedrahimi/ant-design/blob/ant-rtl/components/config-provider/index.en-US.md)
[View rendered components/config-provider/index.zh-CN.md](https://github.com/saeedrahimi/ant-design/blob/ant-rtl/components/config-provider/index.zh-CN.md)

saeedrahimi committed:
Added direction property to ConfigProvider

saeedrahimi committed:
cascader rtl tests added

saeedrahimi committed:
update config-provider doc

saeedrahimi committed:
update config-provider doc

saeedrahimi committed:
RTL: grid system

saeedrahimi committed:
RTL: input component

saeedrahimi committed:
RTL: input component

saeedrahimi committed:
RTL: input component

saeedrahimi committed:
RTL: select component

saeedrahimi committed:
RTL: select component

saeedrahimi committed:
RTL: switch component

saeedrahimi committed:
RTL: switch component

saeedrahimi committed:
RTL: switch component

saeedrahimi committed:
RTL: tree component

saeedrahimi committed:
RTL: tree component

saeedrahimi committed:
fix rtl demo lint

saeedrahimi committed:
fix rtl demo lint

saeedrahimi committed:
fix rtl demo lint

saeedrahimi committed:
rtl demo change en-us description

saeedrahimi committed:
rtl demo change en-us description

saeedrahimi committed:
rtl demo change en-us description

saeedrahimi commented:
I have no idea why lint fails, there is no error on my environment.
Windows 10 v17134.1069 
@afc163 , @zombieJ

saeedrahimi committed:
change bundlesize css limit

zombieJ commented:
`@typescript-eslint/parser` failed with markdown at `2.5.0`. It's OK if not other lint failed without markdown file:

ref https://github.com/typescript-eslint/typescript-eslint/issues/1123

saeedrahimi committed:
RTL: modal component (exclude confirm)

saeedrahimi committed:
RTL: modal component (exclude confirm)

saeedrahimi committed:
RTL: modal component (exclude confirm)

saeedrahimi committed:
RTL: modal component (exclude confirm)

saeedrahimi committed:
RTL: table component

saeedrahimi committed:
RTL: table component

saeedrahimi committed:
RTL: table component

saeedrahimi committed:
RTL: table component

saeedrahimi committed:
RTL: pagination component

saeedrahimi committed:
RTL: pagination component

saeedrahimi committed:
RTL: pagination component

saeedrahimi committed:
RTL: pagination component

saeedrahimi committed:
cleanup rtl demo

saeedrahimi committed:
cleanup rtl demo

saeedrahimi committed:
cleanup rtl demo

saeedrahimi committed:
cleanup rtl demo

mahdi13 commented:
@saeedrahimi I check this page every day! Kindly ask you to finish it.

saeedrahimi committed:
fix pagination.tsx compile error

saeedrahimi committed:
fix pagination.tsx compile error

saeedrahimi committed:
RTL: button and button-group

saeedrahimi committed:
RTL: button and button-group

saeedrahimi committed:
RTL: button and button-group

saeedrahimi committed:
RTL: button and button-group

saeedrahimi committed:
RTL: Steps component

saeedrahimi committed:
RTL: Steps component

saeedrahimi committed:
RTL: Steps component

saeedrahimi committed:
RTL: Steps component

saeedrahimi committed:
fix rtl demo style

saeedrahimi committed:
fix rtl demo style

saeedrahimi committed:
fix rtl demo style

saeedrahimi committed:
fix rtl demo style

saeedrahimi committed:
fix input suffix icon alignment

saeedrahimi committed:
fix input suffix icon alignment

saeedrahimi committed:
fix input suffix icon alignment

zombieJ commented:
@saeedrahimi,
`4.0-prepare` synced with `master`. Some conflict need rebase `4.0-prepare`. Thanks.

saeedrahimi committed:
fix select component arrow issue

saeedrahimi committed:
RTL: form component

saeedrahimi committed:
RTL: form component

saeedrahimi committed:
RTL: form component

saeedrahimi committed:
RTL: form component

saeedrahimi committed:
add pagination rtl test

saeedrahimi committed:
add pagination rtl test

saeedrahimi committed:
fix test lint error

saeedrahimi committed:
RTL: rate component

saeedrahimi committed:
RTL: rate component

saeedrahimi committed:
RTL: rate component

saeedrahimi committed:
RTL: rate component

saeedrahimi committed:
RTL: radio and radio group components

saeedrahimi committed:
RTL: radio and radio group components

saeedrahimi committed:
RTL: radio and radio group components

saeedrahimi committed:
RTL: radio and radio group components

saeedrahimi committed:
RTL: tree-select component

saeedrahimi committed:
RTL: tree-select component

saeedrahimi committed:
RTL: tree-select component

saeedrahimi committed:
RTL: tree-select component

saeedrahimi committed:
some fixes to RTL components

saeedrahimi committed:
RTL: badge component

saeedrahimi committed:
RTL: badge component

saeedrahimi committed:
RTL: badge component

saeedrahimi committed:
RTL: badge component

saeedrahimi committed:
fix rtl issue in inline form

saeedrahimi committed:
fix rtl issue in inline form

saeedrahimi committed:
fix rtl issue in inline form

saeedrahimi committed:
fix rtl issue in inline form

saeedrahimi committed:
fix input component rtl padding issue

saeedrahimi committed:
fix input component rtl padding issue

saeedrahimi committed:
fix input component rtl padding issue

saeedrahimi committed:
fix input component rtl padding issue

saeedrahimi committed:
fix switch component text rtl issue

saeedrahimi committed:
fix table grouped header text-align

saeedrahimi committed:
add rtl support to whole demo with RTL button

saeedrahimi committed:
add rtl support to whole demo with RTL button

saeedrahimi committed:
add rtl support to whole demo with RTL button

saeedrahimi committed:
add rtl support to whole demo with RTL button

saeedrahimi commented:
@afc163 , @zombieJ 
Hi there,
![ant-design-rtl-button](https://user-images.githubusercontent.com/12451743/67944598-0ff6ca00-fbf2-11e9-80e4-7c68aec26318.JPG)

I thought that config-provider demo page is not suitable to demonstrate all `rtl` components and their functionality in rtl direction.
so i added a button to header that changes direction of the whole demo.
I just wanted to know your idea about these changes. is it ok to continue with this?
I wanna do my best.
Thanks.

yoyo837 commented:
Great work, I hope that the owner can pay enough attention to this and like your contribution.

saeedrahimi committed:
Update rtl demo responsive

saeedrahimi committed:
RTL: page-header component

zombieJ commented:
> @afc163 , @zombieJ
> Hi there,
> ![ant-design-rtl-button](https://user-images.githubusercontent.com/12451743/67944598-0ff6ca00-fbf2-11e9-80e4-7c68aec26318.JPG)
> 
> I thought that config-provider demo page is not suitable to demonstrate all `rtl` components and their functionality in rtl direction.
> so i added a button to header that changes direction of the whole demo and updated some styles of `site/theme` to support rtl properly.
> I just wanted to know your idea about these changes. is it ok to continue with this?
> I wanna do my best.
> Thanks.

Cool. I'm OK with this if all the components support RTL.

saeedrahimi committed:
RTL: typography component

saeedrahimi committed:
RTL: Dropdown (Partial)

saeedrahimi committed:
fix lost changes after rebase

saeedrahimi committed:
fix lint errors

saeedrahimi committed:
RTL: Transfer Component

saeedrahimi committed:
RTL: upload component

saeedrahimi committed:
RTL: update avatar demo

saeedrahimi committed:
RTL: comment component

saeedrahimi committed:
RTL: collapse component

saeedrahimi committed:
RTL: carousel component

saeedrahimi committed:
update snapshots

saeedrahimi committed:
RTL: Card component

saeedrahimi committed:
RTL: descriptions component

saeedrahimi committed:
RTL:  Empty component

saeedrahimi committed:
RTL: list component

saeedrahimi committed:
RTL: slider component

saeedrahimi committed:
slider component import/order

yoyo837 commented:
Note that some components have been refactored to a new major version in `4.0-prepare` branch. And there are other components that are also planned to be refactored.

saeedrahimi committed:
add shared rtlTest

saeedrahimi committed:
RTL: Statistic component

saeedrahimi committed:
RTL: tooltip components

saeedrahimi committed:
RTL: popover component

saeedrahimi commented:
> Note that some components have been refactored to a new major version in `4.0-prepare` branch. And there are other components that are also planned to be refactored.

thanks for note, i will double check them to ensure components working correctly.

saeedrahimi committed:
RTL: timeline component

saeedrahimi committed:
RTL: tag component

saeedrahimi committed:
RTL: alert component

saeedrahimi committed:
RTL: drawer component

saeedrahimi committed:
RTL: Tab component

saeedrahimi committed:
change direction definition

saeedrahimi committed:
RTL: progress component

saeedrahimi committed:
input.tsx, remove duplicate after rebase

saeedrahimi committed:
input.tsx, remove duplicate after rebase

saeedrahimi committed:
fix demo.less after rebase

saeedrahimi committed:
fix ant-row-rtl after rebase

saeedrahimi committed:
fix ant-row-rtl after rebase

saeedrahimi committed:
fix upload issues in rtl

saeedrahimi committed:
badge rtl demo margin fix

saeedrahimi committed:
fix: tabs with icon margin

saeedrahimi committed:
fix: radio-wrapper margin

saeedrahimi committed:
fix: table component after rebase

saeedrahimi committed:
fix: centered modal text-align

saeedrahimi committed:
update slider snapshot

saeedrahimi committed:
RTL: popconfirm component

saeedrahimi committed:
RTL: back-top component

saeedrahimi committed:
RTL: spin component

saeedrahimi committed:
RTL: result component

saeedrahimi committed:
RTL: skeleton component

saeedrahimi committed:
RTL: menu component

saeedrahimi committed:
RTL: time-picker component

saeedrahimi committed:
RTL: calendar component

saeedrahimi committed:
RTL: date-picker component

saeedrahimi committed:
RTL: home page

saeedrahimi committed:
update snapshots

saeedrahimi committed:
update snapshots

saeedrahimi committed:
test: add auto-complete rtl test

saeedrahimi committed:
test: add avatar component rtl tests

saeedrahimi committed:
test: add badge component rtl tests

saeedrahimi committed:
test: add breadcrumb component rtl tests

saeedrahimi committed:
test: add button components rtl tests

saeedrahimi committed:
test: add card component rtl tests

saeedrahimi committed:
test: add carousel component rtl tests

saeedrahimi committed:
test: add cascader component rtl tests

saeedrahimi committed:
test: add cascader component rtl tests

saeedrahimi committed:
test: add checkbox component rtl tests

saeedrahimi committed:
test: add collapse component rtl tests

saeedrahimi committed:
test: add comment component rtl tests

saeedrahimi committed:
test: add dropdown component rtl tests

saeedrahimi committed:
test: add empty component rtl tests

saeedrahimi committed:
test: add form component rtl tests

saeedrahimi committed:
test: add grid component rtl tests

saeedrahimi committed:
test: add input component rtl tests

saeedrahimi committed:
test: add search component rtl tests

saeedrahimi committed:
test: add input-number component rtl tests

saeedrahimi committed:
test: add layout component rtl tests

saeedrahimi committed:
test: add list component rtl tests

saeedrahimi committed:
test: add mentions component rtl tests

saeedrahimi committed:
test: add modal component rtl tests

saeedrahimi committed:
test: add page-header component rtl tests

saeedrahimi committed:
test: add pagination component rtl tests

saeedrahimi committed:
test: add pagination component rtl tests

saeedrahimi committed:
test: add radio component rtl tests

saeedrahimi committed:
test: add rate component rtl tests

saeedrahimi committed:
test: add select component rtl tests

saeedrahimi committed:
test: add slider component rtl tests

saeedrahimi committed:
test: add steps component rtl tests

saeedrahimi committed:
test: add switch component rtl tests

saeedrahimi committed:
test: add table component rtl tests

saeedrahimi committed:
test: add transfer component rtl tests

saeedrahimi committed:
test: add tree component rtl tests

saeedrahimi committed:
test: add tree-select component rtl tests

saeedrahimi committed:
test: add typography component rtl tests

saeedrahimi committed:
test: add upload component rtl tests

saeedrahimi committed:
test: add affix component rtl tests

saeedrahimi committed:
update calendar tests

saeedrahimi committed:
update calendar tests

saeedrahimi commented:
@zombieJ , I'm almost done with implementing rtl view of components.
Would you please review this pull request?
( about rtl of `message` and `notification` component i want to open an issue)

saeedrahimi committed:
increase css file maxSize

saeedrahimi committed:
update snapshots

saeedrahimi committed:
update snapshots

saeedrahimi committed:
remove workflows to allow push

saeedrahimi committed:
remove duplicate reverse prop from slider

saeedrahimi committed:
fix: remove table demo from config-provider

saeedrahimi committed:
fix: remove table demo from config-provider

saeedrahimi committed:
fix: remove table demo from config-provider

saeedrahimi committed:
fix: remove table demo from config-provider

saeedrahimi committed:
fix lint error

saeedrahimi committed:
fix lint error

shayegh commented:
سلام
خدا قوت 
انشاالله به زودی نهائی بشه و بتونیم ازش استفاده کنیم

zombieJ commented:
hi @saeedrahimi , v4 plan release beta version on December 1. Is that possible to finish this feature at that time?

saeedrahimi commented:
> @zombieJ , I'm almost done with implementing rtl view of components.
> Would you please review this pull request?
> ( about rtl of `message` and `notification` component i want to open an issue)

> hi @saeedrahimi , v4 plan release beta version on December 1. Is that possible to finish this feature at that time?
Yes,
I'm almost done, for now all components and front-page supports rtl direction. except `message` and `notification` components that i can't access `ConfigContext` in them,

saeedrahimi commented:
> @zombieJ , I'm almost done with implementing rtl view of components.
> Would you please review this pull request?
> ( about rtl of `message` and `notification` component i want to open an issue)
 
> hi @saeedrahimi , v4 plan release beta version on December 1. Is that possible to finish this feature at that time?

Yes,
I'm almost done, for now all components and front-page supports rtl direction. except `message` and `notification` components that i can't access `ConfigContext` in them,

zombieJ commented:
It's OK to leave them currently since it do not have content.

saeedrahimi committed:
update direction.md icons

saeedrahimi commented:
@zombieJ , i'm getting different snapshots in my environment. would you guide me how to resolve failing tests.?
Windows 10
npm 6.10.2

zombieJ commented:
Could you attach the screenshot about error snapshot?

saeedrahimi commented:
> Could you attach the screenshot about error snapshot?

I don't have any error in my environment, all tests are passing successfully in my environment.

saeedrahimi commented:
CI tests are failing because of diff in snapshots. and i don't know why.
I've updated snapshots and still fails.

I checked some of tests, in `ci/circleci: test_dom` :
```

FAIL  components/locale-provider/__tests__/index.test.js (15.979s)
  Locale Provider
...
...
    ✕ should display the text as id (209ms)
...
    mount and unmount
      ✓ component could be updated and unmounted without errors (9ms)

  ● Locale Provider › should display the text as id

    expect(received).toMatchSnapshot()

    Snapshot name: `Locale Provider should display the text as id 1`

    - Snapshot
    + Received

    @@ -331,15 +331,15 @@
                            2017
                          </a>
                        </span>
                        <a
                          class="ant-calendar-next-month-btn"
    -                     title="Bulan depan (PageDown)"
    +                     title="Bulan selanjutnya (PageDown)"
                        />

```
 -   title="Bulan depan (PageDown)" is in  `rc-calendar` version `"9.15.5"` [link](https://github.com/react-component/calendar/blob/d486704af3357741bb150628d206a956a5d1cd61/src/locale/id_ID.js#L21)

+ title="Bulan selanjutnya (PageDown)" is in  `rc-calendar` version `"9.15.7+"` [link](https://github.com/react-component/calendar/blob/0f2c66f618fddf8cd563d3930a77abd8dcce39e5/src/locale/id_ID.js#L21)

and current `rc-calendar` version in `ant-design:4.0-prepare` is "~9.15.5"

yoyo837 commented:
Try delete node_modules and reinstall it.

yoyo837 commented:
![image](https://user-images.githubusercontent.com/6134547/69796876-c1951500-1209-11ea-8897-e9a87e3bdfb1.png)

saeedrahimi committed:
dropdown and cascader default placement in rtl

saeedrahimi committed:
update snapshots

saeedrahimi commented:
After 4 hours of trying hard, finally i could run `npm install` successfully !

p.s: after protests in Iran we got internet black-out, we have internet now but with poor quality!

zombieJ commented:
@saeedrahimi , could you help to check `rc-picker` if should update to support RTL? It's new DatePicker in v4
ref: https://github.com/react-component/picker

zombieJ commented:
> After 4 hours of trying hard, finally i could run `npm install` successfully !
> 
> p.s: after protests in Iran we got internet black-out, we have internet now but with poor quality!

Thanks for your effort!  🌸

saeedrahimi committed:
fix lint errors

saeedrahimi committed:
remove duplicate import

saeedrahimi commented:
> 
> 
> @saeedrahimi , could you help to check `rc-picker` if should update to support RTL? It's new DatePicker in v4
> ref: https://github.com/react-component/picker

Yes, i would, but i can't deliver it before December 1 for v4 release.

saeedrahimi committed:
update snapshots

saeedrahimi committed:
update snapshot

zombieJ commented:
> Yes, i would, but i can't deliver it before December 1 for v4 release.

That's OK. Thanks for your lot lot contribution : )

saeedrahimi committed:
update calendar snapshot

saeedrahimi commented:
Hi @zombieJ ,
Is this pull request planned to merge with v4 release?
I'm done with this pull request and every time base branch changes, the process of resolving conflicts takes long time.

yoyo837 commented:
Let's merge this PR... 🚀

saeedrahimi committed:
update snapshots

saeedrahimi commented:
> Let's merge this PR... 🚀
Conflicts resolved ;)

Note this commit (link)[https://github.com/ant-design/ant-design/pull/19380/commits/3ba281d2e8dd5923eab0c9280ab0b2c024f105b1]
I deleted some workflows to allow push my codes.

yoyo837 commented:
This should be merged before the first beta release, I think.

yoyo837 commented:
Still need you to do a little new work. https://github.com/ant-design/ant-design/pull/20023#issuecomment-564598709

saeedrahimi commented:
> 
> 
> Still need you to do a little new work. [#20023 (comment)](https://github.com/ant-design/ant-design/pull/20023#issuecomment-564598709)

I'm working on it ;)

saeedrahimi commented:
Waiting for [rtl picker to be merged](https://github.com/react-component/picker/pull/6)

saeedrahimi committed:
integrate with new rc-picker

saeedrahimi committed:
update snapshots

saeedrahimi commented:
Added rtl direction to the new date-picker based on  [this pull request](https://github.com/react-component/picker/pull/6).
Just one more step remaining, after [this pull request](https://github.com/react-component/picker/pull/6) merged and published on NPM, update `package.json`

saeedrahimi committed:
fix lint errors

saeedrahimi committed:
update snapshot

saeedrahimi committed:
update snapshots

saeedrahimi commented:
~ All Green :)

zombieJ commented:
Test failed. Please help to check if it's by design:

https://circleci.com/gh/ant-design/ant-design/143653?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

saeedrahimi committed:
update snapshots

zombieJ commented:
`4.0-prepare` update the `rc-select` which may change the snapshot. Please merge and reinstall `node_modules`.

saeedrahimi committed:
update snapshots :|

saeedrahimi commented:
> `4.0-prepare` update the `rc-select` which may change the snapshot. Please merge and reinstall `node_modules`.

I've deleted node_modules folder and reinstalled it. there is problem with `rc-dialog` package.
`4.0-prepare` rc-dialog version is `~7.5.2` and snapshots for modal have `div` with `ant-modal-root` class.
but `<div className={'${prefixCls}-root'}>` has been added in `rc-dialog@7.5.14` [link](https://github.com/react-component/dialog/blob/6be81135e6d0bf7ef459280196a03602049d7bb8/src/Dialog.tsx#L434).

To fix, installed `rc-dialog@7.5.14` locally and updated snapshots, after that reverted `rc-dialog` to `@~7.5.2`

saeedrahimi commented:
All green now,
@yoyo837 @afc163  @zombieJ when this pull request planned to merge?

> I'm done with this pull request and every time base branch changes, the process of resolving conflicts takes long time.

yoyo837 commented:
react-component/tabs/pull/234 block this?

saeedrahimi commented:
> [react-component/tabs/pull/234](https://github.com/react-component/tabs/pull/234) block this?

It could be fixed with another patch that has less maintenance.

zombieJ commented:
> > [react-component/tabs/pull/234](https://github.com/react-component/tabs/pull/234) block this?
> 
> It could be fixed with another patch that has less maintenance.

Should [react-component/tabs/pull/234](https://github.com/react-component/tabs/pull/234) merge or it's out of date?

saeedrahimi commented:
> > > [react-component/tabs/pull/234](https://github.com/react-component/tabs/pull/234) block this?
> > 
> > 
> > It could be fixed with another patch that has less maintenance.
> 
> Should [react-component/tabs/pull/234](https://github.com/react-component/tabs/pull/234) merge or it's out of date?

I've rebased it to make sure it works.
After rebase i tested it with ant-design rtl manually and everything works properly.

zombieJ commented:
> > > > [react-component/tabs/pull/234](https://github.com/react-component/tabs/pull/234) block this?
> > > 
> > > 
> > > It could be fixed with another patch that has less maintenance.
> > 
> > 
> > Should [react-component/tabs/pull/234](https://github.com/react-component/tabs/pull/234) merge or it's out of date?
> 
> I've rebased it to make sure it works.
> After rebase i tested it with ant-design rtl manually and everything works properly.

`+ rc-tabs@10.0.0-alpha.1`

saeedrahimi committed:
update snapshots

saeedrahimi committed:
fix compile error.

saeedrahimi committed:
fix typo after rebase

saeedrahimi committed:
update snapshots

saeedrahimi committed:
remove workflows to allow push

saeedrahimi committed:
update snapshots

saeedrahimi committed:
update snapshots

saeedrahimi committed:
fix dist error

saeedrahimi committed:
front-page css fix

zombieJ commented:
hi @saeedrahimi,
conflict need resolve. Is there any block need I help or we can start to merge prepare.

yoyo837 commented:
He has been waiting for the merger, and has been resolving the conflict after each any change from upstream.This time he has not finished.

saeedrahimi commented:
> hi @saeedrahimi,
> conflict need resolve. Is there any block need I help or we can start to merge prepare.

hi
i'm resolving conflicts. in past 4 hours i rebased twice from `4.0-prepare`.
it takes time to apply 235 commits and resolve their conflicts.

saeedrahimi committed:
update snapshots

saeedrahimi committed:
fix lint and test issues

saeedrahimi committed:
restore cascader index.less

zombieJ commented:
Note: Hold #20570 until this PR merged.

saeedrahimi committed:
update snapshots

zombieJ commented:
@saeedrahimi , ci failed.

zombieJ commented on .github/workflows/deploy-site.yml:
```diff
@@ -1,30 +0,0 @@
-        PUBLISH_BRANCH: gh-pages
-        PUBLISH_DIR: ./_site
-      with:
-        emptyCommits: false
```
This file should not remove

saeedrahimi replied:
There was an error while pushing with these files, github was denying push on workflow.
After some googling fixed by using PersonalAccessToken instead of password for git account.

zombieJ commented:
Some file removed by mistake, should revert back:
* .github/workflows/deploy-site.yml 
* .github/workflows/lighthouse-ci.yml
* .github/workflows/rebase.yml 
* .github/workflows/test.yml

zombieJ commented on components/cascader/index.tsx:
```diff
@@ -95,6 +101,8 @@ export interface CascaderProps {
   /** use this after antd@3.7.0 */
   fieldNames?: FieldNamesType;
+  /** typo props name before antd@3.7.0 */
+  filedNames?: FieldNamesType;
```
This typo has been removed in v4

zombieJ commented on components/cascader/index.tsx:
```diff
@@ -164,7 +172,11 @@ function defaultSortFilteredOption(
+  const { fieldNames, filedNames } = props;
+  if ('filedNames' in props) {
+    return filedNames; // For old compatibility
+  }
```
compatibility code has been removed in v4

zombieJ commented on components/config-provider/__tests__/__snapshots__/components.test.js.snap:
```diff
@@ -4979,7 +4979,7 @@ exports[`ConfigProvider components Card prefixCls 1`] = `
 exports[`ConfigProvider components Carousel configProvider 1`] = `
 <div
-  class="config-carousel"
+  class=""
```
Miss prefix className, should check why.

zombieJ commented on components/config-provider/__tests__/__snapshots__/components.test.js.snap:
```diff
@@ -10850,7 +10850,7 @@ exports[`ConfigProvider components Slider configProvider 1`] = `
   <div>
     <div
-      class="config-tooltip config-slider-tooltip"
+      class="config-tooltip"
```
Miss prefix className, should check why.

saeedrahimi replied:
will be resolved after updating snapshots

zombieJ commented on components/config-provider/__tests__/__snapshots__/components.test.js.snap:
```diff
@@ -10950,7 +10950,7 @@ exports[`ConfigProvider components Slider prefixCls 1`] = `
   <div>
     <div
-      class="prefix-Slider-tooltip prefix-Slider-tooltip"
+      class="prefix-Slider-tooltip"
```
Miss prefix className, should check why.

saeedrahimi replied:
will be resolved after updating snapshots

saeedrahimi replied:
not resolved 🤔

saeedrahimi replied:
that was a mistake, i was looking on outdated changes!

zombieJ commented on components/modal/Modal.tsx:
```diff
@@ -210,13 +211,19 @@ export default class Modal extends React.Component<ModalProps, {}> {
+      { [`${prefixCls}-centered`]: !!centered },
+      wrapClassName,
+      {
+        [`${prefixCls}-wrap-rtl`]: direction === 'rtl',
```
this object can merge with upper object

zombieJ commented on components/popover/__tests__/__snapshots__/index.test.js.snap:
```diff
@@ -1,5 +1,7 @@
 // Jest Snapshot v1, https://goo.gl/fbAQLP
 
+exports[`Popover rtl render component should be rendered correctly in RTL direction 1`] = `<span />`;
```
Seems strange it only render a empty span without rlt content.

saeedrahimi replied:
will be resolved after updating snapshots

zombieJ commented on components/popover/demo/placement.md:
```diff
@@ -27,7 +27,7 @@ const content = (
 
 ReactDOM.render(
-  <div className="demo">
+  <div className="demo" style={{ direction: 'ltr' }}>
```
Do not change origin demo.

saeedrahimi replied:
If i remove this style from demo, it will not displays correctly in rtl direction.

![antd-dir-rtl-in-demo2](https://user-images.githubusercontent.com/12451743/71621397-52ce2180-2be4-11ea-9570-2bdd178ce768.gif)

I can handle it by css with demo id like this:
``` less
.ant-row-rtl #components-tooltip-demo-placement {
    direction: ltr;
}
```
Which way do you prefer?

zombieJ replied:
Prefer additional style instead of inline style : )

zombieJ commented on components/steps/index.tsx:
```diff
@@ -39,14 +40,23 @@ export default class Steps extends React.Component<StepsProps, any> {
+  renderSteps = ({ getPrefixCls, direction }: ConfigConsumerProps) => {
     const prefixCls = getPrefixCls('steps', this.props.prefixCls);
     const iconPrefix = getPrefixCls('', this.props.iconPrefix);
+    const className = classNames({ [`${prefixCls}-rtl`]: direction === 'rtl' });
```
miss origin `className`?

zombieJ commented on components/table/Table.tsx:
```diff
@@ -388,9 +390,12 @@ function Table<RecordType extends object = any>(props: TableProps<RecordType>) {
     spinProps = loading;
   }
 
+  const wrapperClassNames = classNames(`${prefixCls}-wrapper`, className, {
```
`className` is used for inner Table. Not wrapper element.

zombieJ commented on components/tooltip/demo/placement.md:
```diff
@@ -21,7 +21,7 @@ const text = <span>prompt text</span>;
 
 ReactDOM.render(
-  <div className="demo">
+  <div className="demo" style={{ direction: 'ltr' }}>
```
Not modify origin demo.

zombieJ commented on package.json:
```diff
@@ -261,11 +261,11 @@
     {
       "path": "./dist/antd.min.css",
-      "maxSize": "55 kB"
+      "maxSize": "70 kB"
```
Should use latest config

saeedrahimi replied:
my rtl classes increased final css size. so if i decrease `maxSize` ci will fail

zombieJ replied:
👌

saeedrahimi committed:
fix logo in rtl and demo controls

saeedrahimi committed:
ci errors

saeedrahimi committed:
resolve steps/index.tsx conflicts

saeedrahimi committed:
tooltip family demo remove inline style

saeedrahimi committed:
resolve table/Table.tsx conflicts

saeedrahimi committed:
resolve modal/Modal.tsx conflicts

saeedrahimi committed:
resolve cascader/index.tsx conflicts

saeedrahimi committed:
add workflows from upstream

saeedrahimi committed:
update snapshots

zombieJ commented on components/slider/__tests__/__snapshots__/demo.test.js.snap:
```diff
@@ -955,7 +955,7 @@ exports[`renders ./components/slider/demo/show-tooltip.md correctly 1`] = `
   <div>
     <div
-      class="ant-tooltip ant-slider-tooltip"
+      class="ant-tooltip"
```
Miss `ant-slider-tooltip`

saeedrahimi replied:
will be resolved after updating snapshots

saeedrahimi replied:
strange, not resolved 🤔

zombieJ commented on components/dropdown/dropdown.tsx:
```diff
@@ -118,26 +117,42 @@ export default class Dropdown extends React.Component<DropDownProps, any> {
       disabled,
     });
 
+    const overlayClassNameCustomized = classNames(overlayClassName, {
```
should pass origin `overlayClassName` here.

saeedrahimi replied:
Excuse me, I didn't get it. could you clarify this for me?

zombieJ replied:
OPS. wrong file. It's comments here:
https://github.com/ant-design/ant-design/pull/19380/files#diff-186839a30bf8b9d67a4b10bf7c091d5fR261

saeedrahimi replied:
done

zombieJ commented:
Should remove margin-right which break the Home header line:
<img width="1241" alt="截屏2020-01-01下午1 42 50" src="https://user-images.githubusercontent.com/5378891/71638493-b025a980-2c9c-11ea-8e65-6c1736858aa2.png">

zombieJ commented:
Logo should not revert:
<img width="180" alt="截屏2020-01-01下午1 48 41" src="https://user-images.githubusercontent.com/5378891/71638537-76a16e00-2c9d-11ea-98c7-98b9d12af668.png">

saeedrahimi committed:
revert logo to default

saeedrahimi committed:
fix codebox demo direction of placements

saeedrahimi committed:
resolve tooltip overlayClassName conflicts

saeedrahimi committed:
update snapshots

saeedrahimi committed:
update popover test

zombieJ commented on components/cascader/index.tsx:
```diff
@@ -562,7 +586,8 @@ class Cascader extends React.Component<CascaderProps, CascaderState> {
     const getPopupContainer = props.getPopupContainer || getContextPopupContainer;
     const rest = omit(props, ['inputIcon', 'expandIcon', 'loadingIcon']);
-
+    const rcCascaderRtlPopupClassName =
```
missing origin `popupClassName`.

zombieJ commented on components/menu/index.tsx:
```diff
@@ -325,6 +330,7 @@ class InternalMenu extends React.Component<InternalMenuProps, MenuState> {
         value={{
           inlineCollapsed: this.getInlineCollapsed() || false,
           antdMenuTheme: this.props.theme,
+          direction: this.direction,
```
Is that possible to switch `MenuContext.Provider` with `ConfigConsumer` to let `direction` from `ConfigConsumer` to the `MenuContext.Provider`? Since use `this. direction` may failed in concurrent mode.

zombieJ commented on components/select/index.tsx:
```diff
@@ -123,8 +124,10 @@ class Select<ValueType extends SelectValue = SelectValue> extends React.Componen
+      [`${prefixCls}-rtl`]: direction === 'rtl',
     });
-
+    const rcSelectRtlDropDownClassName =
```
should pass origin `dropdownClassName`.

HDaghash commented:
What a great effort @saeedrahimi !,
@zombieJ  many thanks for keep following on this PR I think this should be there ASAP.

Memadaldin commented:
@saeedrahimi Amazing work!

zombieJ commented:
Header button background should be transparent:
<img width="197" alt="截屏2020-01-02上午11 26 18" src="https://user-images.githubusercontent.com/5378891/71650704-c726d300-2d52-11ea-9c93-eb4a78f698fc.png">

ycjcl868 commented:
@saeedrahimi `ConfigProvider` RTL demo border-color should be `#303030`, you can add

```css
<style>
[data-theme="theme"] .direction-components {
  border-bottom: 1px solid #303030;
}
</style>
```


![image](https://user-images.githubusercontent.com/13595509/71651717-c4c87700-2d5a-11ea-951d-2caea9c39d56.png)

shayegh commented:
Hi,
There is an issue in the menu component style. the border of the menu must be at the left side.
please check this one

zombieJ commented:
@saeedrahimi, 
Since to avoid continue conflict, some PR is pending on this PR. I will continue work with your branch by commits. If push conflict, please pull firstly to avoid our code conflict.

zombieJ commented:
@saeedrahimi, 
Since to avoid continue conflict, some PR is pending on this PR. I will continue work with your branch by commits. If push conflict, please pull firstly to avoid our code conflict.

zombieJ commented:
> @saeedrahimi,
> Since to avoid continue conflict, some PR is pending on this PR. I will continue work with your branch by commits. If push conflict, please pull firstly to avoid our code conflict.

To avoid conflict, I will use `merge` instead of `rebase`.

zombieJ committed:
fix: cascader miss popupClassName

zombieJ committed:
fix: fix select missing dropdownClassName

zombieJ committed:
Merge remote-tracking branch 'origin/4.0-prepare' into ant-rtl

zombieJ committed:
chore: Update snapshot

zombieJ committed:
merge origin/prepare

zombieJ committed:
chore: Adjust menu use rtl logic

zombieJ committed:
docs: Update demo line color

saeedrahimi commented:
> Hi,
> There is an issue in the menu component style. the border of the menu must be at the left side.
> please check this one
> ![menu](https://user-images.githubusercontent.com/10929932/71657597-db200400-2d55-11ea-914c-e677bf366521.png)

border is in the left side in `ltr` direction. so it should be in right side in `rtl` direction.
border should not be outside of sidebar

saeedrahimi commented:
Thank you @zombieJ for fixing my mistakes 👍

yoyo837 commented:
Amazing works！

vahidalvandi commented:
thank you how i can use this in nuxt‌?

afc163 commented:
<img width="1290" alt="image" src="https://user-images.githubusercontent.com/507615/71704180-b12d1700-2e13-11ea-8379-8e0f9527df53.png">

Is this header misordered?

saeedrahimi commented:
> <img alt="image" width="1290" src="https://user-images.githubusercontent.com/507615/71704180-b12d1700-2e13-11ea-8379-8e0f9527df53.png">
> 
> Is this header misordered?

I think @zombieJ  fixed it in this [commit](https://github.com/ant-design/ant-design/commit/dafc5a2f1d1fb06901231fba9f45f588a400d40d) cause i see header in correct order in [next.ant.design](https://next.ant.design)

afc163 commented:
Could you help to fix it to correct order?

saeedrahimi commented:
> Could you help to fix it to correct order?

Is there any misordered item? 🤔
![antd-header](https://user-images.githubusercontent.com/12451743/71720041-7a67f880-2e35-11ea-938a-16efb495332a.JPG)

 ZombieJ fixed that.

zombieJ commented:
That's fine. I've resolved.

PejmanNik commented:
Amazing work @vahidalvandi ❤
I just find a bug, the `ant-col-rtl` broken all `:not([class^='ant-col-']):not([class*=' ant-col-'])` selectors

yoyo837 commented:
@PejmanNik Feel free to open a new issue about that.

afsanefda commented:
Hi all,
I have a problem with ant Tabs! **onclick** is not showing the content. **The arrow buttons** are not in the correct position. Here is my code and the implementation result!

```
<ConfigProvider direction="rtl" locale="fa">

        <div className={style.body}>
          <Tabs
            className={style.tab}
            defaultActiveKey="0"
            tabPosition="top"
            style={{ height: 220 }}
          >
            {[...Array(30).keys()].map((i) => (
              <TabPane tab={`تب${i}`} key={i}>
              محتویات
                {i}
              </TabPane>
            ))}
          </Tabs>
        </div>
      </ConfigProvider>
```

![Screenshot (3)](https://user-images.githubusercontent.com/37764909/76705691-0da39100-66f7-11ea-8b4d-3a85ae485659.png)

It would be great if you could help!

afsanefda commented:
Is this working with **Tabs**? Here is my code:

```
<ConfigProvider direction="rtl">
          <Tabs
            className={style.tab}
            defaultActiveKey="0"
            tabPosition="top"
            style={{ height: 220 }}
          >
            {[...Array(30).keys()].map((i) => (
              <TabPane tab={`تب${i}`} key={i}>
                محتویات
                {i}
              </TabPane>
            ))}
          </Tabs>
  </ConfigProvider>
```
and the result is:

![Screenshot (4)](https://user-images.githubusercontent.com/37764909/76705884-c1595080-66f8-11ea-9a5f-b0e68ec2cc95.png)

there is no change for tabs!

adirrapidapi commented:
how to make Sider in right side (rtl)?

xrkffgg commented:
> how to make Sider in right side (rtl)?

You can use https://ant.design/components/config-provider-cn/#components-config-provider-demo-direction to set rtl, so the base Slider can be in right side.
or https://ant.design/components/slider-cn/#components-slider-demo-reverse in ltr

adirrapidapi commented:
@xrkffgg  I added with configProvider the rtl direction but the sider it's still in left side, as you can see in this example [https://stackblitz.com/edit/react-ouxpra-trwwbf?embed=1&file=index.js](https://stackblitz.com/edit/react-ouxpra-trwwbf?embed=1&file=index.js)

adirrapidapi commented:
@xrkffgg any solution?

xrkffgg commented:
> @xrkffgg any solution?

sorry, I just saw. I'll fix it.

mehdiraized commented:
@saeedrahimi i have issue on nextjs
```
Warning: Prop `className` did not match. Server: "ant-badge ant-badge-rtl" Client: "ant-badge"
```

xrkffgg commented:
@mehdiraized Please create new issue https://github.com/ant-design/ant-design/issues/new/choose
